File: src/shared/ui/spec/generalNode.ts
// The GENERAL node — `{ type, props, children }` over the full primitive registry (#3485, slice 3a of
// the keystone #3484).
//
// The existing `KitNode` union covers 8 hardcoded kinds against ~60 primitives in the manifest. This is
// the open form that replaces it (the 8 are deleted in 3c, once the renderer can render this in 3b).
//
// THE VOCABULARY IS DERIVED, NOT DECLARED. Both the set of valid `type`s and the per-type prop rules
// come from `UI_KIT` (`shared/ui/manifest.ts`), so adding a primitive there makes it authorable here
// with no edit to this file. A hand-maintained mirror would drift, and a contract that drifts from the
// code is worse than no contract — agents author against it.
import { UI_KIT, type PrimitiveName, type PropSpec } from "../manifest";
import type { KitNode } from "./schema";

/** A node addressing any primitive in the manifest. `props` is open; the validator constrains it. */
export interface GeneralNode {
  type: PrimitiveName;
  props?: Record<string, unknown>;
  /**
   * Children — general nodes, the legacy kinds (while both vocabularies coexist, 3c), or plain TEXT.
   *
   * Text is included because the manifest types `children` as `node`, and a `node` slot legitimately
   * holds text — `<Text>hello</Text>` is the commonest node in the tree. An array-only type here would
   * be NARROWER than what the validator accepts, and a type that disagrees with the contract it
   * describes is worse than no type: the compiler would reject trees the runtime happily validates.
   */
  children?: Array<GeneralNode | KitNode> | string | number;
}

/** Every primitive name the manifest defines — the closed `type` vocabulary. */
export const PRIMITIVE_NAMES: readonly string[] = UI_KIT.map((p) => p.name);

const BY_NAME = new Map(UI_KIT.map((p) => [p.name as string, p]));

/**
 * What the validator can and cannot enforce, stated exactly (#3485).
 *
 * `PropSpec.type` is a CLOSED 13-member union with an explicit `values` list for enums, so most props
 * get real enforcement. The two genuine gaps are `array` and `object`, whose element/field shapes the
 * manifest keeps in prose (`description`) rather than as a machine-readable schema — so the container
 * type is checked and its CONTENTS are not.
 *
 * This is published rather than left implicit on purpose: `bsc ui validate` reporting "valid" while
 * silently meaning "the names look right" is a half-truth that gets trusted and then bites. Anything
 * that consumes this contract can read exactly where the guarantees stop.
 */
export const VALIDATION_COVERAGE: Readonly<Record<string, "checked" | "container-only">> = {
  string: "checked",
  number: "checked",
  boolean: "checked",
  enum: "checked",
  node: "checked",
  function: "checked",
  color: "checked",
  space: "checked",
  fontSize: "checked",
  tracks: "checked",
  style: "container-only",
  array: "container-only",
  object: "container-only",
};

function isPlainObject(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

/** A node-ish value: a general node (`type`) or a legacy kind node (`kind`). */
function isNodeLike(v: unknown): boolean {
  return isPlainObject(v) && (typeof v.type === "string" || typeof v.kind === "string");
}

/**
 * Check one prop VALUE against its spec. Returns an error message, or null when acceptable.
 *
 * Note `function`: in a data tree a handler can never be a literal — it is the NAME of an action the
 * host exposes (the `bind`/`action` seam). So requiring a string here is not a weaker check than a type
 * check, it is a stricter and more meaningful one: a serialized function would be nonsense, and an
 * inline arrow would mean the tree had stopped being data.
 */
function checkValue(spec: PropSpec, value: unknown): string | null {
  switch (spec.type) {
    case "string":
      return typeof value === "string" ? null : `expected a string`;
    case "number":
      return typeof value === "number" ? null : `expected a number`;
    case "boolean":
      return typeof value === "boolean" ? null : `expected a boolean`;
    case "enum": {
      const allowed = spec.values ?? [];
      if (allowed.length === 0) return null; // an enum with no declared set constrains nothing
      return allowed.includes(value as string)
        ? null
        : `"${String(value)}" is not one of ${allowed.join(", ")}`;
    }
    case "node":
      // A slot: a child node, a list of them, or plain text.
      if (typeof value === "string" || typeof value === "number") return null;
      if (Array.isArray(value)) return value.every(isNodeLike) ? null : `expected nodes or text`;
      return isNodeLike(value) ? null : `expected a node, a list of nodes, or text`;
    case "function":
      return typeof value === "string"
        ? null
        : `expected an action NAME (a string) — a data tree binds handlers by name, it cannot carry a function`;
    case "color":
      return typeof value === "string" ? null : `expected a color string (token or CSS color)`;
    case "space":
    case "fontSize":
      return typeof value === "number" || typeof value === "string"
        ? null
        : `expected a rung name or a number`;
    case "tracks":
      return typeof value === "number" || typeof value === "string"
        ? null
        : `expected a track count (number) or a template string`;
    case "style":
      return isPlainObject(value) ? null : `expected a style object`;
    case "array":
      // Contents deliberately unchecked — see VALIDATION_COVERAGE.
      return Array.isArray(value) ? null : `expected an array`;
    case "object":
      return isPlainObject(value) ? null : `expected an object`;
    default:
      return null;
  }
}

/**
 * Structurally validate a general node tree against the manifest. Returns a flat list of human-readable
 * errors (empty = valid), in the same shape as {@link validateKitNode} so both vocabularies report
 * identically while they coexist.
 *
 * At every node: `type` present and a real primitive · every required prop present · no prop outside
 * the primitive's declared set (UNLESS it declares `passthrough`, which legitimately accepts arbitrary
 * DOM props) · every present prop's value acceptable for its declared `PropType` · recursion into
 * `children` and into any node-valued prop.
 *
 * @param node - the value to validate (typically parsed from agent-emitted JSON)
 * @param path - dotted path used to locate errors (defaults to "$")
 */
export function validateGeneralNode(node: unknown, path = "$"): string[] {
  const errors: string[] = [];
  walkGeneral(node, path, errors);
  return errors;
}

function walkGeneral(node: unknown, path: string, errors: string[]): void {
  if (!isPlainObject(node)) {
    errors.push(`${path}: expected a node object`);
    return;
  }
  const type = node.type;
  if (typeof type !== "string") {
    errors.push(`${path}: missing string "type"`);
    return;
  }
  const spec = BY_NAME.get(type);
  if (!spec) {
    errors.push(`${path}: unknown primitive "${type}"`);
    return;
  }

  const props = node.props;
  if (props !== undefined && !isPlainObject(props)) {
    errors.push(`${path}.props: expected an object`);
    return;
  }
  // Node-level `children` is SUGAR for the `children` PROP. The manifest models children the way React
  // does — a prop of type `node`, declared REQUIRED on containers (Stack/Row/Card/…) — but writing a
  // tree as `props: { children: [...] }` at every level is miserable. Normalising here means both forms
  // validate identically, and a bare `children` is not misread as "missing required prop".
  const given: Record<string, unknown> = { ...((props ?? {}) as Record<string, unknown>) };
  if (node.children !== undefined) given.children = node.children;
  const byProp = new Map(spec.props.map((p) => [p.name, p]));

  // Unknown props — skipped for a passthrough primitive, which forwards arbitrary DOM props
  // (className/style/data-*/handlers) to its root by design. Flagging those would be a false positive.
  if (!spec.passthrough) {
    for (const key of Object.keys(given)) {
      if (!byProp.has(key)) errors.push(`${path}.props.${key}: unknown prop for "${type}"`);
    }
  }

  for (const p of spec.props) {
    const v = given[p.name];
    if (p.required && v == null) {
      errors.push(`${path}.props.${p.name}: missing required prop for "${type}"`);
      continue;
    }
    if (v == null) continue; // an absent optional prop takes the component's default
    const err = checkValue(p, v);
    if (err) errors.push(`${path}.props.${p.name}: ${err}`);
  }

  // Recurse into every node-valued prop (a slot) — `children` included, since it was normalised above.
  // The error path reports where the value was WRITTEN (`$.children[0]` vs `$.props.children[0]`) so a
  // message points at the author's own source rather than at the normalised form.
  const wroteChildrenAtNodeLevel = node.children !== undefined;
  for (const [name, value] of Object.entries(given)) {
    if (byProp.get(name)?.type !== "node") continue;
    const where = name === "children" && wroteChildrenAtNodeLevel ? `${path}.children` : `${path}.props.${name}`;
    if (Array.isArray(value)) {
      value.forEach((child, i) => {
        if (isNodeLike(child)) walkAny(child, `${where}[${i}]`, errors);
      });
    } else if (isNodeLike(value)) {
      walkAny(value, where, errors);
    }
  }
}

/** Dispatch to the right validator while the general form and the legacy kinds coexist (3c removes this). */
function walkAny(node: unknown, path: string, errors: string[]): void {
  if (isPlainObject(node) && typeof node.kind === "string") return; // a legacy node — validateKitNode owns it
  walkGeneral(node, path, errors);
}

File: crates/bsc-ui/src/general_node.rs
//! The GENERAL node validator — `{ type, props, children }` over the full primitive registry (#3485).
//!
//! This is the Rust FACE of the validator whose TypeScript face is `src/shared/ui/spec/generalNode.ts`.
//! Both derive their rules from the SAME generated contract (`src-tauri/data/ui/primitives.json`,
//! emitted from `shared/ui/manifest.ts`), so neither hand-maintains a copy of the vocabulary.
//!
//! ## Why two implementations, and what keeps them honest
//!
//! `bsc ui validate` is a standalone Rust CLI — it cannot import TypeScript — but it is also the
//! surface agents actually author against, so the general form has to be enforced *here* or it is not
//! really enforced at all. Two independent validators is a real risk: drifting apart is worse than
//! having one, because callers trust the agreement precisely where it is most likely to quietly break.
//!
//! So the agreement is TESTED, not asserted: `src-tauri/data/ui/node-validation.fixtures.json` is run
//! by both sides (see [`tests::shared_fixtures_match_the_typescript_validator`] and
//! `nodeValidationFixtures.test.ts`). The error MESSAGES are deliberately identical too — the fixtures
//! assert on message content, so a divergence in wording fails rather than passing quietly.

use serde_json::{Map, Value};

/// The structured prop contract, embedded at compile time. GENERATED from `shared/ui/manifest.ts`
/// (`UPDATE_KITS=1 npx vitest run primitives.gen`), which is why `react-ui.json` is no use here: it
/// flattens each prop's type to a TypeScript type STRING, losing the structured `type` + `values`.
pub const PRIMITIVES_JSON: &str = include_str!("../../../src-tauri/data/ui/primitives.json");

/// The parsed primitive contract, as a name → spec map.
fn primitives() -> Map<String, Value> {
    let parsed: Value = serde_json::from_str(PRIMITIVES_JSON).unwrap_or(Value::Null);
    let mut by_name = Map::new();
    if let Some(list) = parsed.get("primitives").and_then(Value::as_array) {
        for p in list {
            if let Some(name) = p.get("name").and_then(Value::as_str) {
                by_name.insert(name.to_string(), p.clone());
            }
        }
    }
    by_name
}

/// Every primitive name the contract defines — the closed `type` vocabulary.
pub fn primitive_names() -> Vec<String> {
    primitives().keys().cloned().collect()
}

fn is_object(v: &Value) -> bool {
    v.is_object()
}

/// A node-ish value: a general node (`type`) or a legacy kind node (`kind`).
fn is_node_like(v: &Value) -> bool {
    v.get("type").and_then(Value::as_str).is_some() || v.get("kind").and_then(Value::as_str).is_some()
}

/// Check one prop VALUE against its declared type. `None` = acceptable.
///
/// Mirrors `checkValue` in `generalNode.ts` arm for arm, including the wording. Note `function`:
/// in a data tree a handler is the NAME of a host action, never a literal — a stricter and more
/// meaningful rule than a type check, since a tree carrying a function has stopped being data.
fn check_value(ty: &str, values: Option<&Vec<Value>>, v: &Value) -> Option<String> {
    match ty {
        "string" => (!v.is_string()).then(|| "expected a string".to_string()),
        "number" => (!v.is_number()).then(|| "expected a number".to_string()),
        "boolean" => (!v.is_boolean()).then(|| "expected a boolean".to_string()),
        "enum" => {
            let allowed: Vec<&str> = values
                .map(|vs| vs.iter().filter_map(Value::as_str).collect())
                .unwrap_or_default();
            if allowed.is_empty() {
                return None; // an enum with no declared set constrains nothing
            }
            let got = v.as_str().unwrap_or("");
            if allowed.contains(&got) {
                None
            } else {
                let shown = match v {
                    Value::String(s) => s.clone(),
                    other => other.to_string(),
                };
                Some(format!("\"{shown}\" is not one of {}", allowed.join(", ")))
            }
        }
        "node" => {
            if v.is_string() || v.is_number() {
                return None;
            }
            if let Some(arr) = v.as_array() {
                return (!arr.iter().all(is_node_like)).then(|| "expected nodes or text".to_string());
            }
            (!is_node_like(v)).then(|| "expected a node, a list of nodes, or text".to_string())
        }
        "function" => (!v.is_string()).then(|| {
            "expected an action NAME (a string) — a data tree binds handlers by name, it cannot carry a function"
                .to_string()
        }),
        "color" => (!v.is_string()).then(|| "expected a color string (token or CSS color)".to_string()),
        "space" | "fontSize" => (!(v.is_number() || v.is_string()))
            .then(|| "expected a rung name or a number".to_string()),
        "tracks" => (!(v.is_number() || v.is_string()))
            .then(|| "expected a track count (number) or a template string".to_string()),
        "style" => (!is_object(v)).then(|| "expected a style object".to_string()),
        // Contents deliberately unchecked — the manifest keeps array/object shapes in prose, so there
        // is no schema to check against. The container is verified and no further; that gap is stated
        // in `VALIDATION_COVERAGE` rather than left for a caller to discover.
        "array" => (!v.is_array()).then(|| "expected an array".to_string()),
        "object" => (!is_object(v)).then(|| "expected an object".to_string()),
        _ => None,
    }
}

/// Structurally validate a general node tree against the primitive contract. Returns a flat list of
/// human-readable errors (empty = valid), identical to the TypeScript validator's output.
pub fn validate_general_node(node: &Value) -> Vec<String> {
    let mut errors = Vec::new();
    let by_name = primitives();
    walk_general(node, "$", &by_name, &mut errors);
    errors
}

fn walk_general(node: &Value, path: &str, by_name: &Map<String, Value>, errors: &mut Vec<String>) {
    if !is_object(node) {
        errors.push(format!("{path}: expected a node object"));
        return;
    }
    let Some(ty) = node.get("type").and_then(Value::as_str) else {
        errors.push(format!("{path}: missing string \"type\""));
        return;
    };
    let Some(spec) = by_name.get(ty) else {
        errors.push(format!("{path}: unknown primitive \"{ty}\""));
        return;
    };

    let props = node.get("props");
    if let Some(p) = props {
        if !is_object(p) {
            errors.push(format!("{path}.props: expected an object"));
            return;
        }
    }

    // Node-level `children` is SUGAR for the `children` PROP — the contract models children as React
    // does (a `node` prop, required on containers), but authoring `props: { children: … }` at every
    // level is miserable. Normalise so both forms validate identically.
    let mut given: Map<String, Value> =
        props.and_then(Value::as_object).cloned().unwrap_or_default();
    let wrote_children_at_node_level = node.get("children").is_some();
    if let Some(c) = node.get("children") {
        given.insert("children".into(), c.clone());
    }

    let empty = Vec::new();
    let declared = spec.get("props").and_then(Value::as_array).unwrap_or(&empty);
    let passthrough = spec.get("passthrough").and_then(Value::as_bool).unwrap_or(false);

    // Unknown props — skipped for a passthrough primitive, which forwards arbitrary DOM props by
    // design. Flagging those would be a false positive, and a validator that cries wolf gets ignored.
    if !passthrough {
        for key in given.keys() {
            let known = declared
                .iter()
                .any(|d| d.get("name").and_then(Value::as_str) == Some(key.as_str()));
            if !known {
                errors.push(format!("{path}.props.{key}: unknown prop for \"{ty}\""));
            }
        }
    }

    for d in declared {
        let Some(name) = d.get("name").and_then(Value::as_str) else { continue };
        let required = d.get("required").and_then(Value::as_bool).unwrap_or(false);
        let value = given.get(name).filter(|v| !v.is_null());
        let Some(value) = value else {
            if required {
                errors.push(format!("{path}.props.{name}: missing required prop for \"{ty}\""));
            }
            continue; // an absent optional prop takes the component's default
        };
        let d_ty = d.get("type").and_then(Value::as_str).unwrap_or("");
        let values = d.get("values").and_then(Value::as_array);
        if let Some(err) = check_value(d_ty, values, value) {
            errors.push(format!("{path}.props.{name}: {err}"));
        }
    }

    // The node-level actions map (#3496): prop name → host action name. Exists because `passthrough`
    // primitives forward arbitrary DOM props, so their handlers are UNDECLARED and the manifest gives
    // no way to infer them — and inferring from the prop's NAME is exactly what this design avoids.
    if let Some(actions) = node.get("actions") {
        match actions.as_object() {
            None => errors.push(format!("{path}.actions: expected an object of prop → action name")),
            Some(map) => {
                for (prop_name, action_name) in map {
                    let ok = action_name.as_str().map(|s| !s.is_empty()).unwrap_or(false);
                    if !ok {
                        errors.push(format!(
                            "{path}.actions.{prop_name}: expected a non-empty action name (a string)"
                        ));
                    }
                    let declared = declared
                        .iter()
                        .find(|d| d.get("name").and_then(Value::as_str) == Some(prop_name.as_str()));
                    match declared {
                        Some(d) => {
                            let d_ty = d.get("type").and_then(Value::as_str).unwrap_or("");
                            if d_ty != "function" {
                                errors.push(format!(
                                    "{path}.actions.{prop_name}: \"{prop_name}\" is declared as {d_ty}, not a handler, on \"{ty}\""
                                ));
                            }
                        }
                        // Undeclared is legitimate only where undeclared props are (a passthrough
                        // primitive) — that is the whole reason this map exists.
                        None if !passthrough => {
                            errors.push(format!("{path}.actions.{prop_name}: unknown prop for \"{ty}\""));
                        }
                        None => {}
                    }
                }
            }
        }
    }

    // Recurse into every node-valued prop (a slot) — `children` included, since it was normalised.
    // The path reports where the value was WRITTEN, so a message points at the author's own source.
    for (name, value) in &given {
        let is_node_prop = declared.iter().any(|d| {
            d.get("name").and_then(Value::as_str) == Some(name.as_str())
                && d.get("type").and_then(Value::as_str) == Some("node")
        });
        if !is_node_prop {
            continue;
        }
        let where_ = if name == "children" && wrote_children_at_node_level {
            format!("{path}.children")
        } else {
            format!("{path}.props.{name}")
        };
        if let Some(arr) = value.as_array() {
            for (i, child) in arr.iter().enumerate() {
                if is_node_like(child) {
                    walk_any(child, &format!("{where_}[{i}]"), by_name, errors);
                }
            }
        } else if is_node_like(value) {
            walk_any(value, &where_, by_name, errors);
        }
    }
}

/// Dispatch while the general form and the legacy kinds coexist (removed in 3c, #3484).
fn walk_any(node: &Value, path: &str, by_name: &Map<String, Value>, errors: &mut Vec<String>) {
    if node.get("kind").and_then(Value::as_str).is_some() {
        return; // a legacy node — `validate_spec` owns it
    }
    walk_general(node, path, by_name, errors);
}

#[cfg(test)]
mod tests {
    use super::*;

    const FIXTURES_JSON: &str =
        include_str!("../../../src-tauri/data/ui/node-validation.fixtures.json");

    #[test]
    fn the_embedded_contract_parses_and_covers_the_registry() {
        let names = primitive_names();
        assert!(names.len() > 50, "expected the full primitive registry, got {}", names.len());
        for expected in ["Text", "Stack", "Card", "Row", "Button"] {
            assert!(names.iter().any(|n| n == expected), "missing {expected}");
        }
    }

    /// THE cross-language guard. Both validators run these fixtures; the error messages are identical
    /// by construction, so a divergence in wording fails here rather than passing quietly. A case
    /// added to the JSON is automatically enforced on both sides.
    #[test]
    fn shared_fixtures_match_the_typescript_validator() {
        let parsed: Value = serde_json::from_str(FIXTURES_JSON).expect("fixtures are valid JSON");
        let cases = parsed.get("cases").and_then(Value::as_array).expect("cases array");
        assert!(cases.len() > 10, "a silently-empty fixture set would make this guard vacuous");

        for case in cases {
            let name = case.get("name").and_then(Value::as_str).unwrap_or("<unnamed>");
            let node = case.get("node").expect("case has a node");
            let want = case.get("count").and_then(Value::as_u64).unwrap_or(0) as usize;
            let errors = validate_general_node(node);

            assert_eq!(
                errors.len(),
                want,
                "case \"{name}\": expected {want} error(s), got {:?}",
                errors
            );
            if let Some(contains) = case.get("contains").and_then(Value::as_array) {
                for needle in contains.iter().filter_map(Value::as_str) {
                    assert!(
                        errors.iter().any(|e| e.contains(needle)),
                        "case \"{name}\": no error contained \"{needle}\" — got {:?}",
                        errors
                    );
                }
            }
        }
    }

    #[test]
    fn a_passthrough_primitive_does_not_flag_unknown_props() {
        // Guarded separately from the fixtures because the two validators BRANCH on this flag — a
        // wrong value in the generated contract would make them disagree silently.
        let node = serde_json::json!({
            "type": "Text", "children": "x", "props": { "data-testid": "t" }
        });
        assert!(validate_general_node(&node).is_empty());
    }

    #[test]
    fn a_legacy_kind_child_is_left_to_the_legacy_validator() {
        // While both vocabularies coexist, a `kind` node nested under a general one must not be
        // reported as an unknown primitive — it belongs to `validate_spec`.
        let node = serde_json::json!({
            "type": "Stack", "children": [{ "kind": "text", "text": "hi" }]
        });
        assert!(validate_general_node(&node).is_empty(), "{:?}", validate_general_node(&node));
    }
}
